### PR TITLE
Handle local charm resources

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following attempts to clarify what resources should use which
resource client depending on the charm URL.

Previously, before charmhub client was introduced, all charms went
through the use of the local cache (renamed to resourceState) and if it
wasn't there then an attempt to connect to the charmstore client was
made. With the introduction of charmhub this also followed the same 
pattern. Except for local charms, those were never served correctly.
Instead, they were blackholed.

The fix here is to explicitly handle local charms by stating that no
outbound connection is ever used for surfacing a resource. This will
ensure that when deploying a local resource we show the error to the
user in a much more timely manner.

Additionally, in this commit is the re-working/re-naming of all the
indirections to help make more sense. As somebody first coming to
resources and trying to understand the indirections, the naming of these
things were terrible!

"Cache" was used for accessing state, instead, a better word 
"repository" is used to show that it doesn't just serve the data from a 
local cache as it infers by the name.

Naming is hard, but hopefully, this is better for others reading the
code.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy ./acceptancetests/repository/charms/dummy-resource  --resource foo="./acceptancetests/repository/charms/dummy-resource/foo.txt" --resource bar="./acceptancetests/repository/charms/dummy-resource/bar.txt"
$ watch juju status
```

Ensure that the foo/bar resources are deployed.

### CODE REVIEW

File: resource/resourceadapters/opener.go [hunk 1]
```diff
@@ -33,26 +36,30 @@ func newInternalResourceOpener(st ResourceOpenerState, unitName string) (opener
...
-	case charm.CharmHub.Matches(curl.Schema):
-		opener = &ResourceOpener{
-			st:                st,
-			res:               resources,
-			userID:            unit.Tag(),
-			unit:              unit,
-			newResourceOpener: func(st ResourceOpenerState) ResourceRetryClientGetter { return newCharmHubOpener(st) },
+	default:
+		resourceClient = func(st ResourceOpenerState) ResourceRetryClientGetter {
+			return newNopOpener()
```
Comment (SimonRichardson) [inline resource/resourceadapters/opener.go:53]: This is the real fix, everything else is clarity.

File: resource/repositories/cache.go [hunk 2]
```diff
@@ -12,9 +12,9 @@ import (
 	"github.com/juju/juju/resource"
 )
 
-// EntityCache exposes the functionality needed to cache data from
+// EntityRepository exposes the functionality needed to cache data from
```
Comment (manadart) [inline resource/repositories/cache.go:15]: But the file is still _cache.go_ :)

File: resource/resourceadapters/interfaces.go [hunk 3]
```diff
@@ -65,6 +65,8 @@ type Resources interface {
 	SetResource(applicationID, userID string, res charmresource.Resource, r io.Reader) (resource.Resource, error)
 }
 
+type ResourceRetryClientGetterFn func(st ResourceOpenerState) ResourceRetryClientGetter
```
Comment (manadart) [inline resource/resourceadapters/interfaces.go:68]: Needs a comment.

File: resource/resourceadapters/opener.go [hunk 4]
```diff
@@ -108,6 +115,49 @@ func (ro *ResourceOpener) OpenResource(name string) (o resource.Opened, err erro
...
+}
+
+// NewClient opens a new charm store client.
+func (o *nopOpener) NewClient() (*ResourceRetryClient, error) {
+	return newRetryClient(nopClient{}), nil
+}
+
+type nopClient struct{}
+
+func (nopClient) GetResource(req repositories.ResourceRequest) (charmstore.ResourceData, error) {
```
Comment (manadart) [inline resource/resourceadapters/opener.go:157]: Comment here to set the scene of it's usage. I.e. that _not found_ means we can only get it from state.

**Review by manadart — APPROVED** [review]
Looks good and QA fine. I think we need a couple of comments.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**